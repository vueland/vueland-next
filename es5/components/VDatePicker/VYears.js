"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VYears = void 0;

require("../../../src/components/VDatePicker/VYears.scss");

var _vue = require("vue");

var _helpers = require("./helpers");

var _useTransition = require("../../effects/use-transition");

var VYears = (0, _vue.defineComponent)({
  name: 'v-years',
  props: {
    year: [Number, String]
  },
  setup: function setup(props, _ref) {
    var emit = _ref.emit;
    var LIMIT = 100;
    var ON_TABLE = 20;
    var CELLS_IN_ROW = 4;
    var CURRENT_YEAR = new Date().getFullYear();
    var ANIMATION_TIMEOUT = 100;
    var years = (0, _vue.ref)([]);
    var onTableIndex = (0, _vue.ref)(0);
    var isListChanged = (0, _vue.ref)(false);
    var transition = (0, _vue.ref)('');
    var handlers = (0, _vue.inject)('handlers');
    (0, _vue.watchEffect)(function () {
      return isListChanged.value && setTimeout(function () {
        isListChanged.value = false;
      }, ANIMATION_TIMEOUT);
    });
    var computedYear = (0, _vue.computed)({
      get: function get() {
        return +props.year || CURRENT_YEAR;
      },
      set: function set(val) {
        emit('update:year', val);
      }
    });

    if (handlers !== null && handlers !== void 0 && handlers.value) {
      handlers.value = {
        onNext: function onNext() {
          return changeYearsList(true);
        },
        onPrev: function onPrev() {
          return changeYearsList(false);
        }
      };
    }

    function setCurrentTransition(isNext) {
      transition.value = isNext ? 'fade-in-down' : 'fade-in-up';
    }

    function setTableIndex() {
      onTableIndex.value = years.value.findIndex(function (row) {
        return row.find(function (year) {
          return year === computedYear.value;
        });
      });
    }

    function changeYearsList(isNext) {
      var max = years.value.length - 1;
      var val = isNext ? 1 : -1;
      if (onTableIndex.value === max && val > 0 || onTableIndex.value === 0 && val < 0) return;
      setCurrentTransition(isNext);
      onTableIndex.value += val;
      isListChanged.value = true;
    }

    function genTableYears() {
      var fromYear = CURRENT_YEAR - LIMIT;
      var maxYears = LIMIT * 2;
      var yearsList = [];

      for (var i = 0; i <= maxYears; i += 1) {
        if (yearsList.length === ON_TABLE) {
          years.value.push(yearsList);
          yearsList = [];
        }

        yearsList.push(fromYear + i);
      }
    }

    function genYearCell(year) {
      var isSelected = year === computedYear.value;
      var propsData = {
        "class": {
          'v-years__cell': true,
          'v-years__cell--selected': isSelected,
          'v-years__cell--current-year': year === CURRENT_YEAR
        },
        onClick: function onClick() {
          return computedYear.value = year;
        }
      };
      return (0, _vue.h)('div', propsData, year);
    }

    function genYearsRows() {
      var currentYears = years.value[onTableIndex.value];
      var yearsVNodes = currentYears.map(genYearCell);
      return (0, _helpers.genTableRows)(yearsVNodes, 'v-years__row', CELLS_IN_ROW);
    }

    function genYears() {
      var propsData = {
        "class": 'v-years__years'
      };
      return !isListChanged.value && (0, _vue.h)('div', propsData, genYearsRows()) || null;
    }

    genTableYears();
    setTableIndex();
    return function () {
      var content = (0, _useTransition.useTransition)(genYears(), transition.value);
      var propsData = {
        "class": {
          'v-years': true
        }
      };
      return (0, _vue.h)('div', propsData, content);
    };
  }
});
exports.VYears = VYears;
//# sourceMappingURL=VYears.js.map