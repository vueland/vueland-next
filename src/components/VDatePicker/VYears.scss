@import "../../styles/scss/modules/global";
@import "variables";

.v-years {
  width: 100%;
  overflow: hidden;

  &__row {
    @include flex(flex, center, center);
    width: 100%;
  }

  &__cell {
    @include flex(flex, center, center);
    width: 25%;
    padding: 15px 10px;
    cursor: pointer;
    color: currentColor;
    border: thin solid transparent;
    transition: all .2s linear;

    &:hover {
      color: $date-picker-hover-color !important;
    }

    &--selected {
      border-color: currentColor;
    }

    &--current-year {
      border-color: $primary;
    }
  }

}
