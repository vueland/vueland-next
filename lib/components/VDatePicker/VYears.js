import "../../../src/components/VDatePicker/VYears.scss";
import { h, ref, watchEffect, inject, computed, defineComponent } from 'vue';
import { genTableRows } from './helpers';
import { useTransition } from '../../effects/use-transition';
export const VYears = defineComponent({
  name: 'v-years',
  props: {
    year: [Number, String]
  },

  setup(props, {
    emit
  }) {
    const LIMIT = 100;
    const ON_TABLE = 20;
    const CELLS_IN_ROW = 4;
    const CURRENT_YEAR = new Date().getFullYear();
    const ANIMATION_TIMEOUT = 100;
    const years = ref([]);
    const onTableIndex = ref(0);
    const isListChanged = ref(false);
    const transition = ref('');
    const handlers = inject('handlers');
    watchEffect(() => isListChanged.value && setTimeout(() => {
      isListChanged.value = false;
    }, ANIMATION_TIMEOUT));
    const computedYear = computed({
      get() {
        return +props.year || CURRENT_YEAR;
      },

      set(val) {
        emit('update:year', val);
      }

    });

    if (handlers !== null && handlers !== void 0 && handlers.value) {
      handlers.value = {
        onNext: () => changeYearsList(true),
        onPrev: () => changeYearsList(false)
      };
    }

    function setCurrentTransition(isNext) {
      transition.value = isNext ? 'fade-in-down' : 'fade-in-up';
    }

    function setTableIndex() {
      onTableIndex.value = years.value.findIndex(row => {
        return row.find(year => year === computedYear.value);
      });
    }

    function changeYearsList(isNext) {
      const max = years.value.length - 1;
      const val = isNext ? 1 : -1;
      if (onTableIndex.value === max && val > 0 || onTableIndex.value === 0 && val < 0) return;
      setCurrentTransition(isNext);
      onTableIndex.value += val;
      isListChanged.value = true;
    }

    function genTableYears() {
      const fromYear = CURRENT_YEAR - LIMIT;
      const maxYears = LIMIT * 2;
      let yearsList = [];

      for (let i = 0; i <= maxYears; i += 1) {
        if (yearsList.length === ON_TABLE) {
          years.value.push(yearsList);
          yearsList = [];
        }

        yearsList.push(fromYear + i);
      }
    }

    function genYearCell(year) {
      const isSelected = year === computedYear.value;
      const propsData = {
        class: {
          'v-years__cell': true,
          'v-years__cell--selected': isSelected,
          'v-years__cell--current-year': year === CURRENT_YEAR
        },
        onClick: () => computedYear.value = year
      };
      return h('div', propsData, year);
    }

    function genYearsRows() {
      const currentYears = years.value[onTableIndex.value];
      const yearsVNodes = currentYears.map(genYearCell);
      return genTableRows(yearsVNodes, 'v-years__row', CELLS_IN_ROW);
    }

    function genYears() {
      const propsData = {
        class: 'v-years__years'
      };
      return !isListChanged.value && h('div', propsData, genYearsRows()) || null;
    }

    genTableYears();
    setTableIndex();
    return () => {
      const content = useTransition(genYears(), transition.value);
      const propsData = {
        class: {
          'v-years': true
        }
      };
      return h('div', propsData, content);
    };
  }

});
//# sourceMappingURL=VYears.js.map