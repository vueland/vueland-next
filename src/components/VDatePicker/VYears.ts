// Styles
import './VYears.scss'

// VUe API
import { h, ref, watchEffect, inject, computed, defineComponent } from 'vue'

// Helpers
import { genTableRows } from './helpers'

// Types
import { VNode, Ref } from 'vue'
import { DatePickerBtnHandlers } from './VDatepicker'

// Effects
import { useTransition } from '../../effects/use-transition'

const props: any = {
  year: [Number, String],
}

export const VYears = defineComponent({
  name: 'v-years',
  props,

  setup(props, { emit }) {
    // constants
    const LIMIT = 100
    const ON_TABLE = 20
    const CELLS_IN_ROW = 4
    const ANIMATION_TIMEOUT = 100
    const CURRENT_YEAR = new Date().getFullYear()

    // reactive
    const years = ref<Array<number[]>>([])
    const onTableIndex = ref<number>(0)
    const isListChanged = ref<boolean>(false)
    const transition = ref<string>('')

    // injects
    const handlers = inject('handlers') as Ref<DatePickerBtnHandlers>

    watchEffect(
      () => isListChanged.value &&
        setTimeout(() => {
          isListChanged.value = false
        }, ANIMATION_TIMEOUT),
    )

    // computed values
    const computedYear = computed<number>({
      get() {
        return +props.year! || CURRENT_YEAR
      },
      set(val: number) {
        emit('update:year', val)
      },
    })

    // methods
    const setCurrentTransition = isNext => {
      transition.value = isNext ? 'fade-in-down' : 'fade-in-up'
    }

    const setTableIndex = () => {
      onTableIndex.value = years.value.findIndex(row => {
        return row.find(year => year === computedYear.value)
      })
    }

    const changeYearsList = isNext => {
      const max = years.value.length - 1
      const val = isNext ? 1 : -1

      if (
        (onTableIndex.value === max && val > 0) ||
        (onTableIndex.value === 0 && val < 0)
      ) return

      setCurrentTransition(isNext)

      onTableIndex.value += val
      isListChanged.value = true
    }

    const genTableYears = () => {
      const fromYear = CURRENT_YEAR - LIMIT
      const maxYears = LIMIT * 2

      let yearsList: number[] = []

      for (let i = 0; i <= maxYears; i += 1) {
        if (yearsList.length === ON_TABLE) {
          years.value.push(yearsList)
          yearsList = []
        }
        yearsList.push(fromYear + i)
      }
    }

    const genYearCell = year => {
      const isSelected = year === computedYear.value

      const propsData = {
        class: {
          'v-years__cell': true,
          'v-years__cell--selected': isSelected,
          'v-years__cell--current-year': year === CURRENT_YEAR,
        },
        onClick: () => (computedYear.value = year),
      }

      return h('div', propsData, year)
    }

    const genYearsCells = (range): VNode[] => {
      return range.map(genYearCell)
    }

    const genYearsRows = () => {
      const currentYears = years.value[onTableIndex.value]
      const yearsVNodes = genYearsCells(currentYears)

      return genTableRows(
        yearsVNodes,
        'v-years__row',
        CELLS_IN_ROW,
      )
    }

    const genYearsList = (): VNode | boolean => {
      const propsData = {
        class: 'v-years__list',
      }
      return (
        !isListChanged.value && h('div', propsData, genYearsRows())
      )
    }

    const genYearsTable = (): VNode => {
      const content = useTransition(
        genYearsList() as VNode,
        transition.value,
      )
      const propsData = {
        class: {
          'v-years': true,
        },
      }

      return h('div', propsData, content)
    }

    handlers.value = {
      onNext: () => changeYearsList(true),
      onPrev: () => changeYearsList(false),
    }

    genTableYears()
    setTableIndex()

    return () => genYearsTable()
  },
})
