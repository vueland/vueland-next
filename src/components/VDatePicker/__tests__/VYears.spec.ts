import { mount, VueWrapper } from '@vue/test-utils'
import { VYears } from '../VYears'


describe('VYears', () => {
  let mountFunction: (options?: any) => VueWrapper<any>

  beforeEach(() => {
    mountFunction = (options = {}) => mount(VYears, { ...options })
  })

  it('should mount component and match snapshot', () => {
    const cmp = mountFunction()

    expect(cmp.html()).toMatchSnapshot()
  })

  it('should set year prop and match snapshot', () => {
    const cmp = mountFunction({
      props: {
        year: 2020,
      },
    })
    const cells = cmp.findAll('.v-years__cell')

    expect(cmp.props().year).toEqual(2020)
    expect(cells.length).toEqual(20)
    expect(cmp.html()).toMatchSnapshot()
  })

  it('should emit click and select date and match snapshot', async () => {
    const cmp = mountFunction({
      props: {
        year: 2020,
      },
    })

    const cells = cmp.findAll('.v-years__cell')
    await cells[1].trigger('click')

    expect(cells[1].attributes().class).toContain('v-years__cell--selected')
    expect(cmp.html()).toMatchSnapshot()
  })

  it('should test year update event', async () => {
    const stub = jest.fn()
    const cmp = mountFunction({
      props: {
        year: 2020,
        ['onUpdate:year']: stub,
      },
    })

    const cells = cmp.findAll('.v-years__cell')
    await cells[1].trigger('click')

    expect(stub).toBeCalledTimes(1)
    expect(stub).toBeCalledWith(2021)
  })
})
