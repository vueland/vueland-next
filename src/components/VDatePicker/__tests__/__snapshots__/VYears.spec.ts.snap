// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VYears should mount component and match snapshot 1`] = `
<div class="v-date-picker-years">
  <transition-stub>
    <div class="v-date-picker-years__years">
      <div class="v-date-picker-years__row">
        <div class="v-date-picker-years__cell v-date-picker-years__cell--selected v-date-picker-years__cell--current-year">
          2020
        </div>
        <div class="v-date-picker-years__cell">
          2021
        </div>
        <div class="v-date-picker-years__cell">
          2022
        </div>
        <div class="v-date-picker-years__cell">
          2023
        </div>
      </div>
      <div class="v-date-picker-years__row">
        <div class="v-date-picker-years__cell">
          2024
        </div>
        <div class="v-date-picker-years__cell">
          2025
        </div>
        <div class="v-date-picker-years__cell">
          2026
        </div>
        <div class="v-date-picker-years__cell">
          2027
        </div>
      </div>
      <div class="v-date-picker-years__row">
        <div class="v-date-picker-years__cell">
          2028
        </div>
        <div class="v-date-picker-years__cell">
          2029
        </div>
        <div class="v-date-picker-years__cell">
          2030
        </div>
        <div class="v-date-picker-years__cell">
          2031
        </div>
      </div>
      <div class="v-date-picker-years__row">
        <div class="v-date-picker-years__cell">
          2032
        </div>
        <div class="v-date-picker-years__cell">
          2033
        </div>
        <div class="v-date-picker-years__cell">
          2034
        </div>
        <div class="v-date-picker-years__cell">
          2035
        </div>
      </div>
      <div class="v-date-picker-years__row">
        <div class="v-date-picker-years__cell">
          2036
        </div>
        <div class="v-date-picker-years__cell">
          2037
        </div>
        <div class="v-date-picker-years__cell">
          2038
        </div>
        <div class="v-date-picker-years__cell">
          2039
        </div>
      </div>
      <div class="v-date-picker-years__row">
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`VYears should set year prop and match snapshot 1`] = `
<div class="v-date-picker-years">
  <transition-stub>
    <div class="v-date-picker-years__years">
      <div class="v-date-picker-years__row">
        <div class="v-date-picker-years__cell v-date-picker-years__cell--selected v-date-picker-years__cell--current-year">
          2020
        </div>
        <div class="v-date-picker-years__cell">
          2021
        </div>
        <div class="v-date-picker-years__cell">
          2022
        </div>
        <div class="v-date-picker-years__cell">
          2023
        </div>
      </div>
      <div class="v-date-picker-years__row">
        <div class="v-date-picker-years__cell">
          2024
        </div>
        <div class="v-date-picker-years__cell">
          2025
        </div>
        <div class="v-date-picker-years__cell">
          2026
        </div>
        <div class="v-date-picker-years__cell">
          2027
        </div>
      </div>
      <div class="v-date-picker-years__row">
        <div class="v-date-picker-years__cell">
          2028
        </div>
        <div class="v-date-picker-years__cell">
          2029
        </div>
        <div class="v-date-picker-years__cell">
          2030
        </div>
        <div class="v-date-picker-years__cell">
          2031
        </div>
      </div>
      <div class="v-date-picker-years__row">
        <div class="v-date-picker-years__cell">
          2032
        </div>
        <div class="v-date-picker-years__cell">
          2033
        </div>
        <div class="v-date-picker-years__cell">
          2034
        </div>
        <div class="v-date-picker-years__cell">
          2035
        </div>
      </div>
      <div class="v-date-picker-years__row">
        <div class="v-date-picker-years__cell">
          2036
        </div>
        <div class="v-date-picker-years__cell">
          2037
        </div>
        <div class="v-date-picker-years__cell">
          2038
        </div>
        <div class="v-date-picker-years__cell">
          2039
        </div>
      </div>
      <div class="v-date-picker-years__row">
      </div>
    </div>
  </transition-stub>
</div>
`;
